cirq/sim/simulator.py

```diff
@@ -44,7 +44,9 @@ class SimulatesSamples(metaclass=abc.ABCMeta):
     def run(
         self,
         program: Union[circuits.Circuit, schedules.Schedule],
-        param_resolver: Optional[study.ParamResolver] = None,
+        param_resolver: Union[None,
+                              Dict[str, float],
+                              study.ParamResolver] = None,
```

Comment:
I would suggest making a type alias for all the things that a param resolver can be created from, then you can use that in all these other places. For example:
```python
# in resolver.py
class ParamResolver:
    From = Union['ParamResolver', Dict[str, float]]

    # optional: a classmethod to create an instance from any of the union types
    @classmethod
    def from(cls, obj: From) -> 'ParamResolver':
        ...

# in simulator.py, etc.
def run(
    ...,
    param_resolver: Optional[study.ParamResolver.From] = None,
    ...):
```
The name in this example is inspired by Rust's [`From` trait](https://doc.rust-lang.org/std/convert/trait.From.html), but whatever the name I think this it's a good idea to capture these sorts of unions with a type alias. Also, note that the `from` method I've shown can be done in `__new__` as you've done below; my only hesitation with that is that I've seen cases where this logic grows quite complicated for converting from various possible types and having that in `__new__` means you get overhead on _every_ instantiation. That might not be so critical for `ParamResolver`, but it's worth considering.

Reply:
Done. Introduced `cirq.ParamResolverOrSimilarType`.